integrationtests/src/test/java/io/kroxylicious/proxy/ProxyRpcTest.java

```diff
@@ -0,0 +1,154 @@
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import io.kroxylicious.test.ApiMessageSampleGenerator;
+import io.kroxylicious.test.ApiMessageSampleGenerator.ApiAndVersion;
+import io.kroxylicious.test.Request;
+import io.kroxylicious.test.Response;
+import io.kroxylicious.test.client.KafkaClient;
+import io.kroxylicious.test.server.MockServer;
+
+import static io.kroxylicious.proxy.Utils.startProxy;
+import static org.apache.kafka.common.protocol.ApiKeys.API_VERSIONS;
+import static org.apache.kafka.common.protocol.ApiKeys.CONTROLLED_SHUTDOWN;
+import static org.apache.kafka.common.protocol.ApiKeys.SASL_AUTHENTICATE;
+import static org.apache.kafka.common.protocol.ApiKeys.SASL_HANDSHAKE;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class ProxyRpcTest {
+
+    public static final int PROXY_PORT = 9192;
+    public static final String PROXY_ADDRESS = "localhost:" + PROXY_PORT;
+    private static MockServer mockServer;
+    private static KafkaClient kafkaClient;
+    private static KafkaProxy kafkaProxy;
+
+    public record Scenario(String name, Response givenMockResponse, Request whenSendRequest, Request thenMockReceivesRequest, Response thenResponseReceived) {
+    }
+
+    /**
+     * API_VERSIONS is not proxied, kroxy can respond to this itself
+     */
+    private static final Set<ApiKeys> NOT_PROXIED = Set.of(API_VERSIONS);
+
+    @BeforeAll
+    public static void beforeAll() throws InterruptedException {
+        mockServer = MockServer.startOnRandomPort();
+        kafkaClient = new KafkaClient();
+        String config = configure(mockServer);
+        kafkaProxy = startProxy(config);
+    }
+
+    @MethodSource("scenarios")
+    @ParameterizedTest
+    public void testKroxyliciousCanDecodeManipulateAndProxyRPC(Scenario scenario) throws Exception {
+        mockServer.clear();
+        mockServer.setResponse(scenario.givenMockResponse());
+        CompletableFuture<Response> future = kafkaClient.get("localhost", PROXY_PORT, scenario.whenSendRequest());
+        Response response = future.get(10, TimeUnit.SECONDS);
+        assertEquals(scenario.thenMockReceivesRequest(), onlyRequest(mockServer), "unexpected request received at mock for scenario: " + scenario.name());
+        assertEquals(scenario.thenResponseReceived(), response, "unexpected response received from kroxylicious for scenario: " + scenario.name());
+    }
+
+    @AfterAll
+    public static void afterAll() {
+        close(mockServer);
+        close(kafkaClient);
+        close(kafkaProxy);
+    }
+
+    private static void close(AutoCloseable closeable) {
+        try {
+            closeable.close();
+        }
+        catch (Exception e) {
+            e.printStackTrace();
```

Comment:
Will change to logger. I've done this try/catch/log-swallow pattern so that we make a best effort at shutting down the 3 resources. We don't want an exception on mockServer close to prevent closing kroxylicous and probably causing other tests to fail due to the 9192 being in use.

Reply:
Yeah that's a reasonable point about hogging the port but if closing it failed we are probably stuffed anyway...